egs/wsj/s5/steps/cleanup/internal/get_pron_stats.py

```diff
@@ -0,0 +1,203 @@
+# in ctm_prons, we put it in "info" as an entry:  [utt_id, word, nonsil_right]
+# only if it's nonsil_right segment is not empty, which may be used when processing
+# the next word.
+# 
+# Normally, one non-<eps> word is only aligned to one pronounciation candidate. However
+# when there is a preceding/following <eps>, like in the following example, we
+# assume the phones aligned to <eps> should be statistically distributed
+# to its neighboring words (BTW we assume there are no consecutive <eps> within an utterance.)
+# Thus we append the "nonsil_left" segment of these phones to the pronounciation
+# of the preceding word, if the last phone of this pronounciation is not a silence phone,
+# Similarly we can add a pron candidate to the following word.
+# 
+# For example, for the following part of a ctm_prons file:
+# 911Mothers_2010W-0010916-0012901-1 other AH DH ER
+# 911Mothers_2010W-0010916-0012901-1 <eps> K AH N SIL B
+# 911Mothers_2010W-0010916-0012901-1 because IH K HH W AA Z AH
+# 911Mothers_2010W-0010916-0012901-1 <eps> V SIL
+# 911Mothers_2010W-0010916-0012901-1 when W EH N
+# 911Mothers_2010W-0010916-0012901-1 people P IY P AH L
+# 911Mothers_2010W-0010916-0012901-1 <eps> SIL
+# 911Mothers_2010W-0010916-0012901-1 heard HH ER 
+# 911Mothers_2010W-0010916-0012901-1 <eps> D
+# 911Mothers_2010W-0010916-0012901-1 that SIL DH AH T
+# 911Mothers_2010W-0010916-0012901-1 my M AY
+# 
+# The corresponding segment in the "info" list is:
+# [911Mothers_2010W-0010916-0012901-1, other, set('AH DH ER', 'AH DH ER K AH N')]
+# [911Mothers_2010W-0010916-0012901-1, <eps>, 'B'
+# [911Mothers_2010W-0010916-0012901-1, because, set('IH K HH W AA Z AH', 'B IH K HH W AA Z AH', 'IH K HH W AA Z AH V', 'B IH K HH W AA Z AH V')]
+# [911Mothers_2010W-0010916-0012901-1, when, set('W EH N')]
+# [911Mothers_2010W-0010916-0012901-1, people, set('P IY P AH L')]
+# [911Mothers_2010W-0010916-0012901-1, <eps>, 'D']
+# [911Mothers_2010W-0010916-0012901-1, that, set('SIL DH AH T')]
+# [911Mothers_2010W-0010916-0012901-1, my, set('M AY')]
+# 
+# Then we accumulate pronouciation stats from "info". Basically, for each occurence
+# of a word, each pronounciation candidate gets equal soft counts. e.g. In the above
+# example, each pron candidate of "because" gets a count of 1/4. The stats is stored
+# in a dictionary (word, pron) : count.
+
+def GetStatsFromCtmProns(silphones, optional_silence, non_scored_words, ctm_prons_file_handle):
+    info = []
+    for line in ctm_prons_file_handle.readlines():
+        splits = line.strip().split()
+        utt = splits[0]
+        word = splits[1]
+        phones = splits[2:]
+        if phones == []:
+            phones = [optional_silence]
+        # extract the nonsil_left and nonsil_right segments, and then try to
+        # append nonsil_left to the pron candidates of preceding word, getting
+        # extended pron candidates.
+        if word == '<eps>' or (word in non_scored_words and word != '<unk>' and word != '<UNK>'):
```

Comment:
On Aug 27 I wrote in email:
Hi Dan, I noticed that if we allow all non_speech entries in the uncleaned r1 data, in the ctm_pron file, there are many cases like:

KevinStone_2010U-0024782-0025580-1 [UH] EH
KevinStone_2010U-0024782-0025580-1 fda F T
KevinStone_2010U-0024782-0025580-1 [NOISE] IY EY

So I'm thinking I should statistically merge the phones aligned to non_speech words to nearby words, as I did for <eps> before, since they behave like <eps>. Do you agree?

and you replied that makes sense. So I was trying to treat non-scored words (except OOV symbol <unk>) as <eps>. I'll put this example in the script.